flight/PiOS/STM32F30x/pios_ws2811.c

```diff
@@ -0,0 +1,226 @@
+	TIM_TimeBaseStructure.TIM_ClockDivision = 0;
+	TIM_TimeBaseStructure.TIM_CounterMode   = TIM_CounterMode_Up;
+	TIM_TimeBaseInit(ws2811_cfg->timer, &TIM_TimeBaseStructure);
+
+	/* PWM1 Mode configuration: Channel1 */
+	TIM_OCStructInit(&TIM_OCInitStructure);
+	TIM_OCInitStructure.TIM_OCMode      = TIM_OCMode_PWM1;
+	TIM_OCInitStructure.TIM_OutputState = TIM_OutputState_Enable;
+	TIM_OCInitStructure.TIM_Pulse       = 5;
+	TIM_OCInitStructure.TIM_OCPolarity  = TIM_OCPolarity_High;
+
+	switch (ws2811_cfg->timer_chan) {
+	case TIM_Channel_1:
+		TIM_OC1Init(ws2811_cfg->timer, &TIM_OCInitStructure);
+		TIM_OC1PreloadConfig(ws2811_cfg->timer, TIM_OCPreload_Enable);
+		break;
+	case TIM_Channel_2:
+		TIM_OC2Init(ws2811_cfg->timer, &TIM_OCInitStructure);
+		TIM_OC2PreloadConfig(ws2811_cfg->timer, TIM_OCPreload_Enable);
+		break;
+	case TIM_Channel_3:
+		TIM_OC3Init(ws2811_cfg->timer, &TIM_OCInitStructure);
+		TIM_OC3PreloadConfig(ws2811_cfg->timer, TIM_OCPreload_Enable);
+		break;
+	case TIM_Channel_4:
+		TIM_OC4Init(ws2811_cfg->timer, &TIM_OCInitStructure);
+		TIM_OC4PreloadConfig(ws2811_cfg->timer, TIM_OCPreload_Enable);
+		break;
+	}
+
+	TIM_CtrlPWMOutputs(ws2811_cfg->timer, ENABLE);
+
+	DMA_DeInit(ws2811_cfg->dma_chan);
+
+	DMA_StructInit(&DMA_InitStructure);
+	DMA_InitStructure.DMA_PeripheralBaseAddr = (uint32_t)&ws2811_cfg->timer->CCR1 + ws2811_cfg->timer_chan;
+	DMA_InitStructure.DMA_MemoryBaseAddr     = (uint32_t)ws2811_dev->dma_buffer;
+	DMA_InitStructure.DMA_DIR = DMA_DIR_PeripheralDST;
+	DMA_InitStructure.DMA_BufferSize         = ws2811_dev->buffer_size;
+	DMA_InitStructure.DMA_PeripheralInc      = DMA_PeripheralInc_Disable;
+	DMA_InitStructure.DMA_MemoryInc          = DMA_MemoryInc_Enable;
+	DMA_InitStructure.DMA_PeripheralDataSize = DMA_PeripheralDataSize_Word;
+	DMA_InitStructure.DMA_MemoryDataSize     = DMA_MemoryDataSize_Byte;
+	DMA_InitStructure.DMA_Mode     = DMA_Mode_Normal;
+	DMA_InitStructure.DMA_Priority = DMA_Priority_High;
+	DMA_InitStructure.DMA_M2M = DMA_M2M_Disable;
+
+	DMA_Init(ws2811_cfg->dma_chan, &DMA_InitStructure);
+
+	TIM_DMACmd(ws2811_cfg->timer, ws2811_cfg->timer_dma_source, ENABLE);
+
+	TIM_Cmd(ws2811_cfg->timer, ENABLE);
+
+	PIOS_WS2811_set_all(ws2811_dev, 0, 0, 0);
+
+	*dev_out = ws2811_dev;
+
+	return 0;
+}
+
+void PIOS_WS2811_set(ws2811_dev_t dev, int led,
+		uint8_t r, uint8_t g, uint8_t b)
+{
+	if (!dev) {
+		return;
+	}
+
+	if (led >= dev->max_leds) {
+		return;
+	}
+
+	int offset = led * WS2811_BITS_PER_LED;
+
+	uint32_t grb = (g << 16) | (r << 8) | (b);
+
+	for (int bit = (WS2811_BITS_PER_LED - 1); bit >= 0; --bit) {
+		dev->dma_buffer[offset++] =
+				(grb & (1 << bit)) ? BIT_COMPARE_1 : BIT_COMPARE_0;
+	}
+}
+
+void PIOS_WS2811_trigger_update(ws2811_dev_t dev)
+{
+	if (!dev) {
+		return;
+	}
+
+	if (dev->dma_active) {
+		if (!DMA_GetFlagStatus(dev->config->dma_tcif)) {
+			return;
+		}
+
+		dev->dma_active = false;	/* LOL */
+		DMA_ClearFlag(dev->config->dma_tcif);
```

Comment:
Doesn't the if above stop you getting here if that flag is set? This whole thing seems wrong. :/